src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs

```diff
@@ -168,33 +171,38 @@ internal ManifestItem Transform(InternalManifestItem item)
                     var message = $"Error processing {item.LocalPathFromRoot}: {e.Message}";
                     throw new PathTooLongException(message, e);
                 }
-
             }
 
-            if (missingXRefDetails.Count > 0)
+            LogInvalidXRefs(unresolvedXRefs);
+
+            return manifestItem;
+        }
+
+        private void LogInvalidXRefs(List<XRefDetails> unresolvedXRefs)
+        {
+            if (unresolvedXRefs == null || unresolvedXRefs.Count == 0)
             {
-                var distinctUids = string.Join(", ", missingXRefDetails.Select(i => i.RawSource).Distinct().Select(s => $"\"{s}\""));
-                Logger.LogWarning($"Invalid cross reference {distinctUids}.", file: item.LocalPathFromRoot);
-                foreach (var group in missingXRefDetails.GroupBy(i => i.SourceFile))
-                {
-                    int maxCountPerFile = 10;
-                    string details;
+                return;
+            }
 
-                    // For each source file, print the first 10 invalid cross reference
-                    if (group.Count() >= maxCountPerFile)
-                    {
-                        details = $"{string.Join(", ", group.Take(maxCountPerFile).Select(i => $"\"{i.RawSource}\" in line {i.SourceStartLineNumber.ToString()}"))}, etc. Fisrt {maxCountPerFile} invalid cross reference are shown above.";
-                    }
-                    else
-                    {
-                        details = string.Join(", ", group.Select(i => $"\"{i.RawSource}\" in line {i.SourceStartLineNumber.ToString()}"));
-                    }
+            var distinctUids = unresolvedXRefs.Select(i => i.RawSource).Distinct().Select(s => $"\"{HttpUtility.HtmlDecode(s)}\"").ToList();
+            Logger.LogWarning($"{distinctUids.Count} invalid cross reference(s) {distinctUids.ToDelimitedString(", ")}.");
+            foreach (var group in unresolvedXRefs.GroupBy(i => i.SourceFile))
+            {
+                // For each source file, print the first 10 invalid cross reference
+                var details = group.Take(MaxInvalidXrefMessagePerFile).Select(i => $"\"{HttpUtility.HtmlDecode(i.RawSource)}\" in line {i.SourceStartLineNumber.ToString()}").Distinct().ToList();
+                var prefix = details.Count > MaxInvalidXrefMessagePerFile ? $"top {MaxInvalidXrefMessagePerFile} " : string.Empty;
+                var message = $"Details for {prefix}invalid cross reference(s): {details.ToDelimitedString(", ")}";
 
-                    Logger.LogInfo($"Invalid cross reference details: {details}", file: group.Key ?? item.LocalPathFromRoot);
+                if (group.Key != null)
+                {
+                    Logger.LogInfo(message, file: group.Key);
+                }
+                else
+                {
+                    Logger.LogInfo(message);
```

Comment:
>message [](start = 35, length = 7)

if raw source is null, need we keep the logic to log with item.LocalPathFromRoot?